### PR TITLE
typescipt test

### PR DESCRIPTION
**Pull request by colman-hartinger**
Initial Typescript test:

For this route the TS file has to be compiled to JS using `tsc <filename>` that way it can be used in the js files.

I do think there would be some benefits to converting everything to TS especially since I think you mentioned that GrabCADPrint uses typescript but I think that would be a day or twos work and I'm uncertain on the necessity. If we go that route you would use `ts-node cli.js <command>` or compile the TS to JS and use the library normaly

### CODE REVIEW

**Comment by colman-hartinger** [comment]
@bradymadams got the TS conversion working to run the cli you have two options

1: global install
```
npm install -g ts-node
npm install -g typescript
```
then run
```
ts-node cli.ts <command>
```
2: local install
just run npm i after pulling changes then run
```
npx ts-node cli.ts <command>
```

**Comment by colman-hartinger** [comment]
@bradymadams I think this showcases the benefits of typescript a bit more now. All of the data types are defined in object blocks using [interfaces](https://www.typescriptlang.org/docs/handbook/interfaces.html) and each main job object will inherit the properties of the corresponding interface. I still need to set better defaults but as it is now a job object is easily to create and change values and typescript will show an error and won't compile if the wrong datatype is passed in

File: src/job.ts [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+  constructor(config: Model | null = null) {
+    if (config) {
+      this.meshes = config.meshes;
+      this.slicer = config.slicer;
+      this.steps = config.steps;
+    } else {
+      // Set defaults
+      this.meshes = [];
+      this.slicer = {
+        type: 'string',
```
Comment (colman-hartinger) [inline src/job.ts:489]: @bradymadams do you think larger objects like `slicer` should be broken down into their own class?

File: src/job.ts [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+}
+
+export class Job{
+  public type: string;
+  chop: Model;
+  bulk: Material[];
+  extruders: Extruder[];
+  optimization: Optimization;
+
+  constructor(config: Job = null) {
```
Comment (colman-hartinger) [inline src/job.ts:614]: do you think the constructor should take optional parameters i.e. the user can input the chop in the config but the rest is set to the default value?

File: src/job.ts (hunk 1 above)
Comment (bradymadams) [reply src/job.ts:489]: Yes, I do. It looks like you already have a `Slicer` class. What's the reason for not using it here?

Comment (colman-hartinger) [reply src/job.ts:489]: The `Slicer` Class you are talking about is actually an interface which is used to define types within the object. I’ll implement a slicer class which uses that interface

File: src/job.ts (hunk 2 above)
Comment (bradymadams) [reply src/job.ts:614]: Good question. I have a couple thoughts here:

1. I think the argument name should be different if you go with this approach (likewise for all the other constructors in this file). `config` is misleading, IMO.
2. The single `Job` argument is a little confusing to me - how would I get the `Job` object to send to this constructor without starting with an empty Job and building it up manually? In which case, I've already built it up so I don't need to call the Job constructor with my Job object.
3. I think the answer to your question is "it depends". It depends on each class and what the typical use case is. To start out I think it's fine to just have no arguments on the constructors (unless some are needed) and just create empty objects. But I will think about this some more.

File: src/job.ts [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+      this.type = 'validation';
+      this.chop = new Model();
+      this.bulk = [new Material()];
+      this.extruders = [new Extruder()];
+      this.optimization = new Optimization();
+    }
+  } 
+}
+
+class SmartSliceJob extends Job {
```
Comment (bradymadams) [inline src/job.ts:632]: I think this entire class should just be a `toJSON` function in the `Job` class.

Comment (colman-hartinger) [reply src/job.ts:632]: Ya that sounds better, I was unsure if this was cleaner or not

**Pull request closed without merging**